### ISSUE TITLE
A clarification in document

### ISSUE THREAD

**Issue by stkim1** [Redundancy/go-sync#16]
@Redundancy 

Hello Daniel, 
I've recently stumble upon `go-sync` and quite impressed with the underlying implementation and its pairing tests. Test cases easily doubles ( or nearly as much as triples in some cases ) workload and it's a bit discouraging to see them unnoticed. Plus, `go-sync` appears to have broader platform coverage as it stays away from `<netinet/in.h>` and `<arpa/inet.h>`.

A reflection of questions arises after following the code base. 

In [README](https://github.com/Redundancy/go-sync/blame/master/README.md#L13), it is pointed that

> The ZSync mechanism has the weakness that HTTP1.1 ranged requests are not always well supported by CDN providers and ISP proxies. When issues happen, they're very difficult to respond to correctly in software (if possible at all). **Using HTTP 1.0 and fully completed GET requests would be better, if possible.**

This is a very appetizing point as it is not unusual to face such issue from low-end hosting services for unbeknown reasons. Looking into `DoReqeust` of `HTTPBlockSource` where its header is composed, however, we can see the operation requests [`HTTP 1.1` and `Range`](https://github.com/Redundancy/go-sync/blob/master/blocksources/httpblocksource.go#L59-L61)  operation specifically. I thought there would a fallback HTTP 1.0 measure to encounter a possible failure, but it is no avail.

I might not have a comprehensive understanding of how `go-sync` is built. If you could point me a direction where I should turn my head, it would be very much appreciated. 

Thank you very much for this great work!

**Comment by Redundancy** [comment]
You're right - the existing implementation does do something that I explicitly warn against. Note that hypothetically speaking, including cache-busting headers in the responses can help. 

The main reason that I don't do it is that it's quite prescriptive to decide to use HTTP1.0 and come up with an appropriate scheme for chopping up the payload. The ideal situation would often be to match the granularity to your units of change (as described a bit lower down).

However, at least theoretically, it should not be too difficult to provide a different implementation of HttpRequester that could source block ranges from potentially multiple files, and fallback to downloading and caching an entire file if a ranged request failed. The work that would be required would largely be to decide how to chop up the input, lay that out as URIs and if that should be handled as a dynamic concern by a webserver or a static one (say, S3).

In order to achieve this, it would probably be best to embed the Verifier into the requester so that it can handle and respond to more cases internally (See BlockSourceBase).

The intent is that the use of Go's interfaces allows delegation of responsibility - the HttpRequester and BlockSourceBase should be almost entirely pluggable, and how you would get the information about the file splits could be an external concern.

**Comment by stkim1** [comment]
I appreciate your input. I've been digging and starting to draw diagrams to picture how code is constructed. It seems, though, that one of feasible solutions to catch up in time and build the HTTP 1.0 fallback could be writing test cases against possible failures especially related to http connections. 

Coincidentally, you've stated that [ `Network Error handling`](https://github.com/Redundancy/go-sync#network-error-handling---improvement-) is one of the improvements you've sought for, and it seems to be the top priority to production readiness appeared in #15. 

I have had very pleasant experience with network simulator such as [Network Link Conditioner](http://nshipster.com/network-link-conditioner/), albeit only on OSX. I don't believe the very utility could be deployed for such test cases, but the notion of having artificial conditioner to stimulate network issues should stand valid for some degree.

In search of tools for such purpose, I've encountered [tylertreat/comcast](https://github.com/tylertreat/comcast) and [Shopify/toxiproxy](https://github.com/Shopify/toxiproxy). 

Both projects are highly oriented for CLI input, but i think it could be possible to incorporate one of them into tests within affordable time window. I'll start building test cases, and make PRs. I hope you can save some time to go over the results, and help me cover corners properly.
